### PR TITLE
Configuration files

### PR DESCRIPTION
**Pull request by Kney-Delach**
Added configuration writer and parser (reader). Note that the way in which these vpa files can be edited should be done through the correct encoding, for example, using Qt Creator (which automatically sets the encoding properties).

### CODE REVIEW

File: Vulkan-Pipeline-Assistant/shaderanalytics.h [hunk 1]
```diff
@@ -15,6 +15,32 @@
...
+
+        bool operator==(const SpirvResource& other) {
+            return name == other.name && size == other.size && resourceType == other.resourceType;
+        }
+        bool operator!=(const SpirvResource& other) {
+            return !operator==(other);
+        }
+    };
+
+
```
Comment (Qozul) [inline Vulkan-Pipeline-Assistant/shaderanalytics.h]: Don't need to add in the enums and structs, they are defined in spirvresource.h. You only need the line 19 forward declaration, remove excess code.

File: Vulkan-Pipeline-Assistant/vulkanmain.cpp [hunk 2]
```diff
@@ -53,11 +53,17 @@ void VulkanMain::CreateVkInstance() {
         qFatal("Could not create Vulkan Instance %d", m_instance.errorCode());
     }
 }
-
+//todo: merge this -- Configuration Branch --
```
Comment (Qozul) [inline Vulkan-Pipeline-Assistant/vulkanmain.cpp]: Remove unneeded todo

File: Vulkan-Pipeline-Assistant/vulkanmain.h [hunk 3]
```diff
@@ -21,6 +21,7 @@ namespace vpa {
         VulkanMain(QWidget* parent);
         ~VulkanMain();
 
+        //todo: merge this -- Configuration Branch --
```
Comment (Qozul) [inline Vulkan-Pipeline-Assistant/vulkanmain.h:24]: Remove unneeded todo

File: Vulkan-Pipeline-Assistant/shaderanalytics.cpp [hunk 4]
```diff
@@ -109,6 +113,38 @@ void ShaderAnalytics::CreateModule(ShaderStage stage, const QString& name) {
...
+        break;
+    }
+    case ShaderStage::GEOMETRY:
+    {
+        m_pConfig->writablePipelineConfig.geomShaderBlob = blob;
+        break;
+    }
+    default:
+        assert(false);
+    }
```
Comment (Qozul) [inline Vulkan-Pipeline-Assistant/shaderanalytics.cpp:147]: Using a switch is fine. Remove todo and curly braces for case scopes (unneeded), curly brace line 119 should be on same line as switch.

File: Vulkan-Pipeline-Assistant/PipelineConfig.h [hunk 5]
```diff
@@ -21,39 +25,33 @@ namespace vpa {
...
 
-    struct PipelineConfig {
-        // Shaders
-        char* vertShader = nullptr;
-        char* fragShader = nullptr;
-        char* tescShader = nullptr;
-        char* teseShader = nullptr;
-        char* geomShader = nullptr;
+    struct WritablePipelineConfig
+    {
```
Comment (Qozul) [inline Vulkan-Pipeline-Assistant/PipelineConfig.h]: Put opening curly brace on existing line (line 28) as with other code.

File: Vulkan-Pipeline-Assistant/filemanager.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                file.close();
+                return;
+            }
+            buffer.erase(std::remove(buffer.begin(), buffer.end(), '\n'), buffer.end());
+            size = buffer.size();
+            buffer.shrink_to_fit();
+            if(!object.LoadConfiguration(buffer, size))
+            {
+                qCritical("Failed to read configuration, check the file format!");
+            }
```
Comment (Qozul) [inline Vulkan-Pipeline-Assistant/filemanager.h]: Put opening curly braces on existing line as with other code.

File: Vulkan-Pipeline-Assistant/pipelineconfig.cpp [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+#define FLOAT_DECIMAL_PLACES 8
+
+template <typename T>
+std::string FloatToString(const T a_value, const int n = FLOAT_DECIMAL_PLACES)
+{
+    std::ostringstream out;
+    out.precision(n);
+    out << std::fixed << a_value;
+    return out.str();
+}
```
Comment (Qozul) [inline Vulkan-Pipeline-Assistant/pipelineconfig.cpp:23]: Indentation

File: Vulkan-Pipeline-Assistant/pipelineconfig.cpp [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+        activeData = GetNextLine(&readPosition, &buffer);
+        if(!(activeData == "VPA_CONFIG_END" || activeData == "\rVPA_CONFIG_END"))
+        {
+            qCritical("VPA_CONFIG_END format invalid, missing object footer.");
+            return false;
+        }
+        return true;
+    }
+
+}
```
Comment (Qozul) [inline Vulkan-Pipeline-Assistant/pipelineconfig.cpp:339]: Put opening curly braces on existing line as with other code.

**Review by Qozul — CHANGES_REQUESTED** [review]
Looks very good overall, address comments then merge.

**Review by Kney-Delach — COMMENTED** [review]
Updated files as per @Qozul review. Almost removed unneeded headers.

**Review by Qozul — APPROVED** [review]
Looks good

**Pull request merged**